Group configs

jagerman opened the pull request:
- Add new user groups config for tracking communities and legacy closed groups (and new closed groups in the future).
- Updated group terminology to match new Session terms: API bits named along the lines of "legacy closed group" are now "legacy group", and "open group" is now "community".
- Added new fields to contacts:
    - `hidden` (true = conversation hidden)
    - `priority` (higher = pinned higher; 0 = unpinned)
    - `exp_mode` expiration mode (named enum, defaulting to expiration_mode::none)
    - `exp_timer` the expiration timer, in minutes
- user groups similarly get `priority` for both communities/legacy groups, and legacy groups also get `hidden` and `disappearing_timer`.
- Abstracted the general "community" related settings (like base url, room, pubkey) into a new `community` class as it is shared between convo volatile and the new user groups configs.  (This also contains the general purpose static methods for community URL and name parsing/validation).
- profile picture `keylen` is removed; pic keys must now always be 32 bytes.
- the `pubkey` argument is gone from the `get_community` (previously `get_open`) method; we now will know the pubkey from the record if the record is found.
- `get_or_construct_community` still takes the pubkey, and the *given argument* pubkey becomes the pubkey in the object, even if what is stored is different.  (What is stored will be updated when the record is `set()`).
- fixed a bug where the outer service node entry (base URL key, plus "#" containing the pubkey) would remain in the config forever even after all rooms from that server were removed.
- exports the version (fixes #14)

jagerman committed:
Enforce profile pic key length of 32 bytes

jagerman committed:
Add user groups plus many small changes

- Add new user groups config for tracking communities and legacy closed
  groups (and new closed groups in the future).
- Updated group terminology to match new Session terms: API bits named
  along the lines of "legacy closed group" are now "legacy group", and
  "open group" is now "community".
- Added new fields to contacts:
    - `hidden` (true = conversation hidden)
    - `priority` (higher = pinned higher; 0 = unpinned)
    - `exp_mode` expiration mode (named enum, defaulting to
      expiration_mode::none)
    - `exp_timer` the expiration timer, in minutes
- user groups similarly get `priority` for both communities/legacy
  groups, and legacy groups also get `hidden` and `disappearing_timer`.
- Abstracted the general "community" related settings (like base url,
  room, pubkey) into a new `community` class as it is shared between
  convo volatile and the new user groups configs.  (This also contains
  the general purpose static methods for community URL and name
  parsing/validation).
- profile picture `keylen` is removed; pic keys must now always be 32
  bytes.
- the `pubkey` argument is gone from the `get_community` (previously
  `get_open`) method; we now will know the pubkey from the record if the
  record is found.
- `get_or_construct_community` still takes the pubkey, and the *given
  argument* pubkey becomes the pubkey in the object, even if what is
  stored is different.  (What is stored will be updated when the record
  is `set()`).
- fixed a bug where the outer service node entry (base URL key, plus "#"
  containing the pubkey) would remain in the config forever even after
  all rooms from that server were removed.

jagerman committed:
Fix dirty-but-immutable state bug

If we merge configs that triggers a merge message, we would get in a
state where we *don't* have a mutable config message, but still need
things to be marked dirty (because of the merge), which is an invalid
state.

This fixes it: if such a case occurs we now convert the ConfigMessage
into a MutableConfigMessage (*without* incrementing its seqno, since
that is already done as part of the merge), which leaves everything in a
proper dirty state *and* lets us make more changes (if needed) to push
the merge + new changes all at once.

Includes a test case that reproduces it when run without the fixes made here.

jagerman committed:
Avoid redundant move warning

mpretty-cyro commented on include/session/config/contacts.h:
```diff
@@ -19,6 +20,12 @@ typedef struct contacts_contact {
+
+    int priority;
+
+    CONVO_EXPIRATION_MODE exp_mode;
```
I get a build error here on iOS `Must use 'enum' tag to refer to type 'CONVO_EXPIRATION_MODE` - updating it to the below works though
```
enum CONVO_EXPIRATION_MODE exp_mode;
```

jagerman replied:
Fixed (with a `typedef` in the expiring.h header so that the `enum` isn't needed here or elsewhere).

jagerman committed:
Simplify struct string handling

Currently the string_view optimization in returned fields creates
awkwardness in the API where callers have to worry about lifetime of the
instance and then we have to provide extra backing strings for cases
where the user doesn't have a lifetime.

This causes bugs, both on the client side from dangling views, and
causes issues where, for instance, the default copy and move
constructors are broken (because we copy/move the string, but not the
string_view).  Rather than fix those constructors, I simplified the API
to be less error prone.

Detailed changes:

- data structs now own their own data, i.e. there are only
  strings/ustrings, but no longer any stored string_views/ustring_views.
  The latter can be slightly more optimal but introduces a bunch of API
  complications and gotchas that the tiny optimization isn't worthwhile.

- contact info name/nickname are now just strings instead of
  optional<string>s.  We now just treat an empty string for either as
  the "not set" value.

- profile_pic now embed its own "optional" state (i.e. url and/or key
  not set, or key length wrong), and so the places we had an
  `optional<profile_pic>` now have a concrete `profile_pic`.  The test
  for "do we have a profile pic" is still the same (that is: `if
  (profile_pic) { ...  }`) but accessing its url/key are now direct `.`
  instead of needing to deference via `->`.

- Some methods for setting things where we had the
  dual-string/string_view behaviour are now gone; you can now just set
  the properties directly.  For example it's now safe to just do
  `contact.name = somename;` instead of needing
  `contact.set_name(somename);` to do the save-a-local-copy behaviour.
  (This is going to break the current code a little bit but should be an
  easy fix).

- There are some changed methods now that do a set-and-validate for
  fields that have specific requirements; they throw when setting
  something valid.  They are optional in many cases: you can just set
  things directly, but if you set something invalid (e.g. non-32-byte
  key) then it'll just get ignored when processing the record.

- the C APIs changed a little to go along with this: rather than storing
  just char* we now have sized `char` arrays to hold the actual string
  content.
  - One important-to-note effect of this is that the profile_pic `.key`
    field is now only usable if there is a non-empty `.url` set
    (otherwise it will just be undefined content).

- Because of the C API fixed size buffers, we also now need to impose
  limits on the relevant fields: name, nickname,
  have some changes as well to go along with this; some
  important things to note:
  - `user_profile_pic.key` should now only be used if `.url` is a non-empty string

jagerman committed:
Expose version

jagerman committed:
C typedef dance for CONVO_EXPIRATION_MODE enum

mpretty-cyro commented on include/session/version.h:
```diff
@@ -0,0 +1,19 @@
+extern const char* LIBSESSION_UTIL_VERSION_FULL;
+
+/// Just the version component as a string, e.g. `v0.1.2-release`.
+extern const char* LIBSESSION_UTIL_VERSION_STR;
```
When trying to access this from Swift I'm getting a "Undefined symbol" error (it's not strictly needed but I was going to include it in the logs in case it helps debugging down the line)

jagerman replied:
Should be fixed now.

mpretty-cyro replied:
Yep, can confirm this is fixed 👍

majestrate commented on src/CMakeLists.txt:
```diff
@@ -40,10 +42,44 @@ target_link_libraries(config
+        configure_file("${CMAKE_CURRENT_SOURCE_DIR}/version.c.in" "${CMAKE_CURRENT_BINARY_DIR}/version.c")
+    endif()
+endif()
+add_library(version version.c)
```
`version.c` to `"${CMAKE_CURRENT_BINARY_DIR}/version.c"`

may fix morgan's link error issue.

jagerman replied:
No, that wasn't it (the `version` lib was missing from the combined archive).

jagerman committed:
Add user groups tests (& fixes)

jagerman committed:
Fix version linkage and add simple version test

jagerman committed:
Hack around ubuntu gcc false positive warning

jagerman committed:
Expose max length variables in C API

majestrate commented on include/session/config/profile_pic.h:
```diff
@@ -6,14 +6,16 @@ extern "C" {
-    // profile pic.
-    const char* url;
-    // The profile pic decryption key, in bytes.  This is a byte buffer of length `keylen`, *not* a
-    // null-terminated C string.  Will be NULL if there is no profile pic.
-    const unsigned char* key;
-    size_t keylen;
+    // Null-terminated C string containing the uploaded URL of the pic.  Will be length 0 if there
+    // is no profile pic.
+    char url[224];
```
i feel like the size of `url` and the value of `PROFILE_PIC_MAX_URL_LENGTH` will very easily get out of sync here. if that happens that feels like a very easy to footcannon regression.

majestrate replied:
suggestion:

make `PROFILE_PIC_MAX_URL_SIZE` a macro (`_PROFILE_PIC_MAX_SIZE_VAL`) and have the size of `url` depend on it, `char url[1+_PROFILE_PIC_MAX_SIZE_VAL]`

then have `extern const size_t PROFILE_PIC_MAX_SIZE_URL` set to `_PROFILE_PIC_MAX_SIZE_VAL` in the compilation unit.

jagerman replied:
It cannot get out of sync because there's a static_assert that will fail to compile if it does.

majestrate commented on include/session/config/profile_pic.hpp:
```diff
@@ -1,39 +1,57 @@
-  private:
-    std::string url_;
-    ustring key_;
+    static constexpr size_t MAX_URL_LENGTH = 223;
```
given changes from c header are applied, then this can be:

`static constexpr size_t MAX_URL_LENGTH = _PROFILE_PIC_MAX_URL_SIZE_VAL;`

mpretty-cyro commented on include/session/config/community.hpp:
```diff
@@ -0,0 +1,239 @@
+    // The returned room name is *not* normalized, that is, it preserve case.
+    //
+    // Throw std::invalid_argument if anything in the URL is unparseable or invalid.
+    static std::tuple<std::string, std::string, ustring> parse_full_url(std::string_view full_url);
```
Can you please add a C interface for parsing a community URL? (figure it would be better to replace the custom Swift function with the libSession one)

jagerman replied:
Added (in `session/config/community.h`)

jagerman replied:
It also has one now to produce a URL from the parts.

mpretty-cyro commented on tests/test_config_user_groups.cpp:
```diff
@@ -0,0 +1,369 @@
+    CHECK(c.enc_seckey.empty());
+    CHECK(c.priority == 0);
+    CHECK(c.name == "");
+    CHECK(c.members().empty());
```
Is there a C interface for the user groups legacy group `members`?

jagerman replied:
Will open a new PR with the additions.

mpretty-cyro commented on tests/test_config_user_groups.cpp:
```diff
@@ -0,0 +1,369 @@
+
+    session::config::UserGroups g2{seed, groups.dump()};
+    CHECK_FALSE(groups.needs_push());
+    CHECK_FALSE(groups.needs_dump());
```
Is this line redundant? (Have the same thing with a comment 2 lines down)

jagerman committed:
Fix segfault

God knows how this was working at all.

jagerman committed:
Change disappearing timer to seconds instead of minutes

mpretty-cyro commented on tests/test_config_user_groups.cpp:
```diff
@@ -0,0 +1,369 @@
+    CHECK(x2->room() == "SudokuRoom");  // Case preserved from the stored value, not the input value
+    CHECK(x2->room_norm() == "sudokuroom");
+    CHECK(x2->pubkey_hex() == to_hex(open_group_pubkey));
+    CHECK(x2->priority == 14);
```
@jagerman I seem to be getting a `priority` of `0` when retrieving the community via the C library here (I don't think I'm doing anything different...)

mpretty-cyro replied:
The legacy group priority check seems to be working fine though (`CHECK(c1.priority == 3)`)

jagerman replied:
Should be fixed in cefe8c2

jagerman committed:
Fix C API not setting priority properly

jagerman committed:
Add note-to-self priority to user profile config

jagerman committed:
Add full url-taking versions or get_community

These just parse the url and then call the decomposed argument version.

jagerman committed:
Add full_url construction

This adds calls to make a full_url either from a community-derived type
(like convo_info_volatile_community or community_info) to construct an
"official" full URL representation of the community URL.

It also adds this into the C API, along with exposing the ability to
parse a full URL into its canonical form parts.
